src/LayerEditor/ui/panels/surfaces.py

```diff
@@ -501,76 +384,88 @@ def _set_message(self, text):
-        assert self.data._idx == idx
+        assert self.data.get_index_from_list(self.le_model.surfaces_model.surface) == idx
         assert idx >= 0
-        assert idx < len(self.layers.surfaces)
-        if not self.layers.delete_surface(idx):
+        assert idx < len(self.le_model.surfaces_model.surfaces)
+        if not self.le_model.delete_surface(idx):
             err_dialog = GMErrorDialog(self)
             err_dialog.open_error_dialog("Surface in use. Can not remove it.")
             return None
 
         # propose new idx
-        new_idx = min(idx, len(self.layers.surfaces) - 1)
-        self.data = SurfFormData.init_from_surface(self.layers.surfaces[new_idx], new_idx)
-        self._fill_forms()
+        new_idx = min(idx, len(self.le_model.surfaces_model.surfaces) - 1)
+        self.data = SurfaceItem(self.e_model.surfaces_model.surfaces[new_idx], new_idx)
+        self.update_forms()
         self.show_grid.emit(self.wg_view_button.isChecked())
 
-
-
-
     def add_surface_from_file(self):
         """
         Handle wg_add_button. Open a file and add new unsaved surface into combo and current panel content.
         """
         if not self.empty_forms():
             return
-        new_data = SurfFormData()
+        new_data = SurfaceItem()
         new_data.file_skip_lines = self.data.file_skip_lines
         new_data.file_delimiter = self.data.file_delimiter
         data = self._load_file(new_data)
         if data:
-            new_data.set_name_from_file([surf.name for surf in self.layers.surfaces])
+            if not self.le_model.surfaces_model.surfaces:
+                quad = new_data.get_actual_quad()
+                points = [QtCore.QPointF(point[0], -point[1]) for point in quad]
+                points.append(QtCore.QPointF(quad[0][0], -quad[0][1]))
+                rect = QtGui.QPolygonF(points).boundingRect()
+                x_adjust = rect.width() / 6
+                y_adjust = rect.height() / 6
+                rect.adjust(-x_adjust, -y_adjust, x_adjust, y_adjust)
+                self.first_surface_added.emit(rect)
+            new_data.set_name_from_file(self.le_model)
             self.data = new_data
             self.wg_view_button.setChecked(True)
-        self._fill_forms()
-
+        self.update_forms()
+        self.show_grid.emit(self.wg_view_button.isChecked())
 
     def reload(self):
         self._load_file(self.data)
         self.wg_view_button.setChecked(True)
-        self._fill_forms()
+        self.update_forms()
 
     def _load_file(self, surface_data):
         # save layer data first
-        if cfg.curr_file is None:
+        if self.le_model.curr_file is None:
             QtWidgets.QMessageBox.information(
                 self, 'Save layer data',
                 'Layer data file must be save first.')
-            cfg.main_window._layer_editor.save_file()
-        if cfg.curr_file is None:
+            self.mainwindow._layer_editor.save_file()
```

Comment:
OK, probably the seva_file method should not be part of the main window, or we can just pass that method in.

Reply:
Similar for the le_model.curr_file. There should probably be a file management class.

Reply:
Save file is not part of main_window and it must be part of top level object since its saving everything. curr_model might not be correct name, its just the origin file from which this le_model was constructed/to which it was saved last.